### PR TITLE
Previously Selected R Version is now being correctly displayed on Choose R Modal

### PR DESCRIPTION
**Pull request by matheustavaresdev**
### Intent

To show the user the correct option for the last R Version that was previously selected when the `Choose R` modal opens again.

### Approach

Saving that path and comparing with the options on the modal to correctly pre-select the current one.

### Automated Tests

None

### QA Notes

Refer to #11156

### Checklist

- [X] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [X] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [X] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [X] This PR passes all local unit tests

Addresses #11156

### CODE REVIEW

File: src/node/desktop/src/ui/widgets/choose-r/preload.ts [hunk 1]
```diff
@@ -132,3 +169,9 @@ const callbacks: Callbacks = {
 };
 
 contextBridge.exposeInMainWorld('callbacks', callbacks);
+
+function isRVersionSelected(selectedVersion: string, versionToCompare: string) {
+  return (
+    selectedVersion.replace(/\//g, '/').replace(/\\/g, '/') === versionToCompare.replace(/\//g, '/').replace(/\\/g, '/')
```
Comment (gtritchie) [inline src/node/desktop/src/ui/widgets/choose-r/preload.ts:175]: Create and use utility function instead of repeating these regexprs over and over.

**Review by gtritchie — CHANGES_REQUESTED** [review]

File: src/node/desktop/src/main/preferences/electron-desktop-options.ts [hunk 2]
```diff
@@ -272,7 +273,7 @@ export class DesktopOptionsImpl implements DesktopOptions {
     if (process.platform !== 'win32') {
       return;
     }
-    this.config.set(kRBinDir, rBinDir);
+    this.config.set(kRBinDir, rBinDir.replace(/\//g, '/').replace(/\\/g, '/'));
```
[the remark this answers is not in the document]
Comment (gtritchie) [reply src/node/desktop/src/main/preferences/electron-desktop-options.ts:276]: Be sure to include unit tests for the function(s) you create.

File: src/node/desktop/src/ui/widgets/choose-r/preload.ts [hunk 3]
```diff
@@ -73,9 +84,16 @@ ipcRenderer.on('initialize', (event, data) => {
...
   });
 
+  const selectWidget = document.getElementById('select') as HTMLSelectElement;
+  selectWidget.disabled = true;
+
   rInstalls.forEach((rInstall) => {
     // normalize separators, etc
-    rInstall = path.normalize(rInstall).replace(/[/\\]+$/g, '');
+    rInstall = path
+      .normalize(rInstall)
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/ui/widgets/choose-r/preload.ts]: Done!

File: src/node/desktop/src/main/preferences/electron-desktop-options.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/main/preferences/electron-desktop-options.ts]: Done!

File: src/node/desktop/src/main/preferences/electron-desktop-options.ts [hunk 4]
```diff
@@ -291,6 +292,30 @@ export class DesktopOptionsImpl implements DesktopOptions {
     return rBinDir;
   }
 
+  // Windows-only option
+  public setRstudioPath(rstudioPath: string): void {
+    if (process.platform !== 'win32') {
+      return;
+    }
+    console.log('setting rstudio path to ',rstudioPath.replace(/\//g, '/').replace(/\\/g, '/'));
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/main/preferences/electron-desktop-options.ts]: Done!

File: src/node/desktop/src/ui/widgets/choose-r/preload.ts (hunk 1 above)
Comment (matheustavaresdev) [reply src/node/desktop/src/ui/widgets/choose-r/preload.ts]: Done!

File: src/node/desktop/src/core/file-path.ts [hunk 5]
```diff
@@ -43,6 +43,11 @@ function normalizeSeparatorsNative(path: string) {
   return normalizeSeparators(path, separator);
 }
 
+// Makes sure that the path only has `/` instead of `\` or `//`
+export const threatPathString = (path: string) => {
```
Comment (kevinushey) [inline src/node/desktop/src/core/file-path.ts:47]: I don't quote understand what this function is doing. It seems identical to https://github.com/rstudio/rstudio/pull/11168/files?w=1#diff-e763e5885fa9463dcc1490ef5c031c7eb737975eace1eecf59f4694e6ca82276R37-R39, except it doesn't let you choose the separator?

I also don't understand the name selected.

File: src/node/desktop/src/main/detect-r.ts [hunk 6]
```diff
@@ -113,6 +112,7 @@ export async function promptUserForR(platform = process.platform): Promise<Expec
     // save the stored version of R
     const path = data.binaryPath as string;
     ElectronDesktopOptions().setRBinDir(dirname(path));
+    ElectronDesktopOptions().setRstudioPath(path);
```
Comment (kevinushey) [inline src/node/desktop/src/main/detect-r.ts:115]: This is the path to R, not RStudio. Shouldn't this be something like `setRBinaryExecutablePath()`?

In addition, if we're setting this, it seems like `rBinDir` is redundant, or at least the executable path could be computed from the binary path. In other words, I don't see why this is necessary.

File: src/node/desktop/src/ui/utils.ts [hunk 7]
```diff
@@ -40,3 +41,8 @@ export const checkForNewLanguage = () => {
     }, 100);
   });
 };
+
+// Makes sure that the path only has `/` instead of `\` or `//`
+export const threatPathString = (path: string) => {
```
Comment (kevinushey) [inline src/node/desktop/src/ui/utils.ts:46]: This looks like a duplicate definition. Is it needed?

File: src/node/desktop/src/main/preferences/electron-desktop-options.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/main/preferences/electron-desktop-options.ts]: Done!

File: src/node/desktop/src/core/file-path.ts (hunk 5 above)
Comment (matheustavaresdev) [reply src/node/desktop/src/core/file-path.ts:47]: I was afraid of messing with the `normalizeSeparators` or `normalizeSeparatorsNative`. 

In my tests for solving this particular issue, only removing any kind of `\\`, `\` over to `/` would work as expected. 

So I thought of using this standalone method. I think that we could even remove `normalizeSeparatorsNative` and use only `normalizeSeparators`, with the same Regexp you mentioned below. Altough I was afraid of breaking something and having to spend more time by testing out stuff. What are your thoughts on it?

File: src/node/desktop/src/main/detect-r.ts (hunk 6 above)
Comment (matheustavaresdev) [reply src/node/desktop/src/main/detect-r.ts:115]: I followed the naming used by `rBinDir` as `rstudioPath` was also declared there and it worked. Not sure how `rBinDir` is being used right now. That's why I created that new one for saving the exact R binary executable path so we could compare on the Choose R Modal to correctly check the current set option. Just let me know if you think I should rename something or do it differently so I can deploy a change!

File: src/node/desktop/src/core/file-path.ts [hunk 8]
```diff
@@ -43,6 +43,11 @@ function normalizeSeparatorsNative(path: string) {
...
+  return path.replace(/\//g, '/').replace(/\\/g, '/');
```
[the remark this answers is not in the document]
Comment (kevinushey) [reply src/node/desktop/src/core/file-path.ts:48]: But there's no need to replace `/` with `/`, so in your above implementation the only thing really happening is `\` is being transformed to `\`.

File: src/node/desktop/src/main/detect-r.ts (hunk 6 above)
Comment (kevinushey) [reply src/node/desktop/src/main/detect-r.ts:115]: Perhaps we should replace the `RBinDir` preference and replace it with the `RExecutable` preference, and compute the `RBinDir` from that preference where appropriate. It's easier to compute the binary directory from the executable path anyhow, I think.

File: src/node/desktop/src/ui/widgets/choose-r/preload.ts [hunk 9]
```diff
@@ -132,3 +166,7 @@ const callbacks: Callbacks = {
 };
 
 contextBridge.exposeInMainWorld('callbacks', callbacks);
+
+function isRVersionSelected(selectedVersion: string, versionToCompare: string) {
+  return threatPathString(selectedVersion) === threatPathString(versionToCompare);
```
Comment (kevinushey) [inline src/node/desktop/src/ui/widgets/choose-r/preload.ts:171]: If I understand correctly, you want to compare the paths after normalization. If that's the case, I think using `normalizeSeparators()` is appropriate here.

File: src/node/desktop/src/core/file-path.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/core/file-path.ts:48]: I just reused `.replace(/\//g, '/')` due for this [line](https://github.com/rstudio/rstudio/blob/main/src/node/desktop/src/main/gwt-callback.ts#L407) being used before, so as I didn't really know what was it for I ended up keeping it haha

File: src/node/desktop/src/ui/widgets/choose-r/preload.ts (hunk 9 above)
Comment (matheustavaresdev) [reply src/node/desktop/src/ui/widgets/choose-r/preload.ts:171]: I just want to make sure both are equal, with no `\`

File: src/node/desktop/src/core/file-path.ts [hunk 10]
```diff
@@ -35,10 +35,10 @@ const homePathAlias = '~/';
 const homePathLeafAlias = '~';
 
 function normalizeSeparators(path: string, separator = '/') {
-  return path.replace(/[\\/]/g, separator);
+  return path.replace(/[\\/]+/g, separator);
 }
 
-function normalizeSeparatorsNative(path: string) {
+export function normalizeSeparatorsNative(path: string) {
   const separator = process.platform === 'win32' ? '\\' : '/';
```
Comment (gtritchie) [inline src/node/desktop/src/core/file-path.ts:42]: Use Node's constant for path separators. `path.sep`

https://nodejs.org/dist/latest-v16.x/docs/api/path.html#pathsep

File: src/node/desktop/src/ui/utils.ts [hunk 11]
```diff
@@ -40,3 +41,13 @@ export const checkForNewLanguage = () => {
     }, 100);
   });
 };
+
+// Makes sure that the path only has `/` instead of `\` or `\\` or `//`
+function normalizeSeparators(path: string, separator = '/') {
```
Comment (gtritchie) [inline src/node/desktop/src/ui/utils.ts:46]: Why does this function (and the next one) exist in both utils.ts and file-path.ts?

File: src/node/desktop/test/unit/core/file-path.test.ts [hunk 12]
```diff
@@ -603,6 +603,23 @@ describe('FilePath', () => {
       assert.strictEqual(cPath.completeChildPath('../bar'), cPath);
       assert.strictEqual(cPath.completeChildPath('/path/to/quux'), cPath);
     });
+    it('Paths only contain forward slashes with no duplicates', () => {
+      const paths = ['c:\\www\\app\\my/folder/file.r',
+        'C:\\R\\4.1.2\\bin\\\\R.exe',
+        'c:\\\\www\\\\app\\my/folder/file.r'];
+      paths.forEach((path) => {
+        assert.isTrue(path.includes('\\'), `Path ${path} should contain at least a single backward slash for this test to be valid`);
+        const threatedPath = normalizeSeparatorsNative(path);
```
Comment (gtritchie) [inline src/node/desktop/test/unit/core/file-path.test.ts:612]: Not sure what this variable name means, specificially the use of "threat".

File: src/node/desktop/src/core/file-path.ts [hunk 13]
```diff
@@ -35,10 +35,10 @@ const homePathAlias = '~/';
 const homePathLeafAlias = '~';
 
 function normalizeSeparators(path: string, separator = '/') {
```
Comment (gtritchie) [inline src/node/desktop/src/core/file-path.ts:37]: Add a documentation comment to this method describing what it does, /** */.

**Review by gtritchie — CHANGES_REQUESTED** [review]

**Review by gtritchie — CHANGES_REQUESTED** [review]

File: src/node/desktop/test/unit/core/file-path.test.ts [hunk 14]
```diff
@@ -603,6 +603,23 @@ describe('FilePath', () => {
       assert.strictEqual(cPath.completeChildPath('../bar'), cPath);
       assert.strictEqual(cPath.completeChildPath('/path/to/quux'), cPath);
     });
+    it('Paths only contain forward slashes with no duplicates', () => {
+      const paths = ['c:\\www\\app\\my/folder/file.r',
+        'C:\\R\\4.1.2\\bin\\\\R.exe',
+        'c:\\\\www\\\\app\\my/folder/file.r'];
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/test/unit/core/file-path.test.ts:609]: Thanks for the reminder. That's true. I will also check this up.

File: src/node/desktop/test/unit/core/file-path.test.ts [hunk 15]
```diff
@@ -603,6 +603,23 @@ describe('FilePath', () => {
...
+      paths.forEach((path) => {
+        assert.isTrue(path.includes('\\'), `Path ${path} should contain at least a single backward slash for this test to be valid`);
+        const threatedPath = normalizeSeparatorsNative(path);
+        assert.isFalse(
+          threatedPath.includes('\\'),
+          `Path ${threatedPath} should NOT contain backward slashes for this test to be valid`,
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/test/unit/core/file-path.test.ts:615]: I'm gonna double check this as well.

File: src/node/desktop/src/ui/utils.ts [hunk 16]
```diff
@@ -40,3 +41,13 @@ export const checkForNewLanguage = () => {
     }, 100);
   });
 };
+
+// Makes sure that the path only has `/` instead of `\` or `\\` or `//`
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/ui/utils.ts:45]: Ok

File: src/node/desktop/src/core/file-path.ts (hunk 13 above)
Comment (matheustavaresdev) [reply src/node/desktop/src/core/file-path.ts:37]: Alright

File: src/node/desktop/test/unit/core/file-path.test.ts (hunk 12 above)
Comment (matheustavaresdev) [reply src/node/desktop/test/unit/core/file-path.test.ts:612]: I meant treat! Gonna also fix this.

File: src/node/desktop/src/ui/utils.ts (hunk 11 above)
Comment (matheustavaresdev) [reply src/node/desktop/src/ui/utils.ts:46]: As it's being used in the `UI` folder, sometimes code from other parts of the code are not available at some point, like in some standalone pages. Kinda like another module. Had some issues with it previously, but also realized that for this part, there are some code being used outside the `ui` folder. So I'm gonna test removing that and leaving in only one place.

File: src/node/desktop/src/core/file-path.ts (hunk 10 above)
Comment (matheustavaresdev) [reply src/node/desktop/src/core/file-path.ts:42]: Alright

**Comment by matheustavaresdev** [comment]
# This is still in progress. Will re-request review when it's ready again.

File: src/node/desktop/src/ui/utils.ts (hunk 11 above)
Comment (matheustavaresdev) [reply src/node/desktop/src/ui/utils.ts:54]: `renderer` folder cannot access some node packages. As `file-path.ts do contain some, it's best to have a different file for functions that are strictly non-node dependent

**Comment by matheustavaresdev** [comment]
This is now ready for review again. cc @gtritchie @kevinushey

**Comment by matheustavaresdev** [comment]
@gtritchie , after testing network R versions, the app cannot load `R.exe` files from the network. Altough, if we mount the network drive, it will work just as it should ( example: `T:\R-3.6.3\bin\x64\R.exe` )

That's the error if we try to access directly a network file. Not sure if this was working previously on QT app. Will try it out.
image

**Comment by matheustavaresdev** [comment]
After tests, it does work on regular QT app. I think there might be something wrong with the current implementation on Windows. Should an issue be opened for this?

cc @kevinushey @gtritchie 

image

**Comment by matheustavaresdev** [comment]
Also, on electron, we select the `R.exe` directly and after selected it opens the app already. 

On QT we select the `folder`, then the `R.exe` is added to the list of options so we can choose it manually and only then open the app. This behavior is not occuring on Electron.

**Comment by matheustavaresdev** [comment]
Let me take a look.

**Comment by kevinushey** [comment]
I'm referring to your comment in https://github.com/rstudio/rstudio/pull/11168#issuecomment-1128408415.

**Comment by matheustavaresdev** [comment]
Oh Gotcha! Yea. I still haven't taken a look on that, but do you think that this issue should be part of this issue or should we open another issue for this error?

**Comment by kevinushey** [comment]
Let's open a separate issue for that.

File: src/node/desktop/src/main/preferences/electron-desktop-options.ts [hunk 17]
```diff
@@ -268,27 +271,42 @@ export class DesktopOptionsImpl implements DesktopOptions {
   }
 
   // Windows-only option
-  public setRBinDir(rBinDir: string): void {
+  public rBinDir(): string {
+    if (process.platform !== 'win32') {
+      return '';
+    }
+
+    let rBinDir: string = dirname(this.rExecutablePath());
```
Comment (kevinushey) [inline src/node/desktop/src/main/preferences/electron-desktop-options.ts:279]: Will this do the right thing if `rExecutablePath()` is not set, or is the empty string?

```
$ node -p 'require("path").dirname("")'
.
```

I think you need to check the value of  `this.rExecutablePath()` before using it.

File: src/node/desktop/src/ui/utils.ts [hunk 18]
```diff
@@ -1,3 +1,5 @@
+import path from "path";
```
Comment (kevinushey) [inline src/node/desktop/src/ui/utils.ts:1]: Is this import still needed? The comment below suggests `path` cannot be used here.

Comment (matheustavaresdev) [reply src/node/desktop/src/ui/utils.ts:1]: Removing! Thanks for noticing. Forgot to remove from there

File: src/node/desktop/src/main/preferences/electron-desktop-options.ts (hunk 17 above)
Comment (matheustavaresdev) [reply src/node/desktop/src/main/preferences/electron-desktop-options.ts]: Added this check on [fc3a811](https://github.com/rstudio/rstudio/pull/11168/commits/fc3a8110b86fb07c1bf205513b80fa921b242af7)

File: src/node/desktop/src/ui/utils.ts [hunk 19]
```diff
@@ -40,3 +43,27 @@ export const checkForNewLanguage = () => {
...
+ * Removes duplicated separators from a path based on platform.
+ *
+ * @export
+ * @param {string} _path
+ * @return {*} 
+ */
+export function normalizeSeparatorsNative(_path: string) {
+  /* using conditional to set the separator based on platform as `path` is not available here */
+  const separator = process.platform === 'win32' ? '\\' : '/';
+  return normalizeSeparators(_path, separator);
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/ui/utils.ts:66]: Fixed

File: src/node/desktop/src/ui/utils.ts [hunk 20]
```diff
@@ -40,3 +43,27 @@ export const checkForNewLanguage = () => {
...
+}
+
+/**
+ * Removes duplicated separators from a path based on platform.
+ *
+ * @export
+ * @param {string} _path
+ * @return {*} 
+ */
+export function normalizeSeparatorsNative(_path: string) {
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/ui/utils.ts:63]: Fixed

File: src/node/desktop/src/ui/utils.ts [hunk 21]
```diff
@@ -40,3 +43,27 @@ export const checkForNewLanguage = () => {
...
+ */
+function normalizeSeparators(path: string, separator = '/') {
+  return path.replace(/[\\/]+/g, separator);
+}
+
+/**
+ * Removes duplicated separators from a path based on platform.
+ *
+ * @export
+ * @param {string} _path
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/ui/utils.ts:60]: Fixed

File: src/node/desktop/src/main/preferences/electron-desktop-options.ts [hunk 22]
```diff
@@ -268,27 +271,42 @@ export class DesktopOptionsImpl implements DesktopOptions {
...
 
   // Windows-only option
-  public rBinDir(): string {
+  public rExecutablePath(): string {
     if (process.platform !== 'win32') {
       return '';
     }
 
-    let rBinDir: string = this.config.get(kRBinDir, properties.platform.default.windows.rBinDir);
+    const rExecutablePath: string = this.config.get(kRExecutablePath, properties.platform.default.windows.rstudioPath);
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/main/preferences/electron-desktop-options.ts:307]: Fixed

File: src/node/desktop/src/main/preferences/electron-desktop-options.ts [hunk 23]
```diff
@@ -39,7 +42,7 @@ const kRendererEngine = 'renderer.engine';
 const kRendererUseGpuExclusionList = 'renderer.useGpuExclusionList';
 const kRendererUseGpuDriverBugWorkarounds = 'renderer.useGpuDriverBugWorkarounds';
 
-const kRBinDir = 'platform.windows.rBinDir';
+const kRExecutablePath = 'platform.windows.rstudioPath';
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/main/preferences/electron-desktop-options.ts:45]: Fixed

File: src/node/desktop/src/core/file-path.ts [hunk 24]
```diff
@@ -34,13 +34,27 @@ export class FilePathError extends Error {
...
+/**
+ * Removes duplicated separators from a path based on platform.
+ *
+ * @export
+ * @param {string} _path
+ * @return {*} 
+ */
+export function normalizeSeparatorsNative(_path: string) {
+  const separator = path.sep;
+  return normalizeSeparators(_path, separator);
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/core/file-path.ts:57]: Fixed

File: src/node/desktop/src/core/file-path.ts [hunk 25]
```diff
@@ -34,13 +34,27 @@ export class FilePathError extends Error {
...
 }
 
-function normalizeSeparatorsNative(path: string) {
-  const separator = process.platform === 'win32' ? '\\' : '/';
-  return normalizeSeparators(path, separator);
+/**
+ * Removes duplicated separators from a path based on platform.
+ *
+ * @export
+ * @param {string} _path
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/core/file-path.ts:52]: Fixed

File: src/node/desktop/src/core/file-path.ts [hunk 26]
```diff
@@ -34,13 +34,27 @@ export class FilePathError extends Error {
...
+ * @return {*} 
+ */
+export function normalizeSeparatorsNative(_path: string) {
```
[the remark this answers is not in the document]
Comment (matheustavaresdev) [reply src/node/desktop/src/core/file-path.ts:55]: Fixed

**Review by kevinushey — APPROVED** [review]
LGTM!

@gtritchie, do you want to handle the issues related to UNC / network paths in this PR, or separately?

**Comment by gtritchie** [comment]
> @gtritchie, do you want to handle the issues related to UNC / network paths in this PR, or separately?

Separately. We do want to fix it but isn't high priority.

**Review by gtritchie — APPROVED** [review]

**Pull request merged**